### ISSUE TITLE
8bit quantisation

### ISSUE THREAD

**Issue by pfeatherstone** [ultralytics/yolov5#211]
Do you plan on supporting quantized models using https://pytorch.org/docs/stable/quantization.html?
That would be really great to further reduce inference speed.

**Comment by glenn-jocher** [comment]
Yes, we already use quantization in our iDetection app for example.

Quantization is appropriate when exporting a pretrained model, though I don't think there's utility to attempting to use this within the repository. Training is already done with automatic mixed precision (amp), and inference is already done automatically at FP16 on GPU (FP32 on CPU, as CPU operations do not support lower precision).

Also do not assume that quantization leads to improved speeds. iDetection FPS is the same regardless of 32, 16 or 8 bit quantized model deployment. The main difference is the model size, which allows consumer apps to benefit from smaller app bundle sizes and faster download speeds/install speeds.

**Comment by pfeatherstone** [comment]
Oh ok fair enough. I thought maybe uint8 models would be a bit faster because integer arithmetic is a lot faster than floating point arithmetic. Maybe the added overhead of avoiding integer overflows cancels out the performance gain of integer operations. Maybe I need to read more about it

**Comment by glenn-jocher** [comment]
@pfeatherstone I think the key to obtaining speed improvements at different precisions is the capability of the hardware to support it. For example the newer nvidia volta GPUs have extra tensor cores dedicated to FP16 operations, so for example running an FP16 model on a T4 GPU will speed it up 2X vs running the same model at FP32. If you try this trick on a 1080Ti or P100 GPU however, you get the same exact speed at FP16/32.

**Comment by pfeatherstone** [comment]
A CPU supports integers just fine though. I’m really liking yolov5s because it’s quite fast on CPU. Interested to see how far one can take it.

**Comment by glenn-jocher** [comment]
@pfeatherstone hmm interesting. Well, if you can figure out how to run CPU inference after quantizing yolov5s or converting it to FP16 please let us know, that might be quite useful for people with limited hardware.

**Comment by tino926** [comment]
> Also do not assume that quantization leads to improved speeds. iDetection FPS is the same regardless of 32, 16 or 8 bit quantized model deployment. The main difference is the model size, which allows consumer apps to benefit from smaller app bundle sizes and faster download speeds/install speeds.

I have different experience with quantization. 
I have converted my yolov5 trained model to .onnx then to openvino, and then convert some nodes of the openvino model to int8. With openvino 2021, converting model from float32 to int 8 increases the fps from ~39 to ~51. (test on intel cpu) (no fps improvement if with openvino 2020)
Currently I have to manually select the nodes to quantize to avoid too much accuracy reduction. I hope there is an easier way to quantize yolov5. It's just worth it now.

**Comment by muetazabdulsamad-oss** [comment]
Hi, I am using YOLOv5n and exporting it to fully INT8 TFLite to run on Google Coral (EdgeTPU).

However, after full INT8 quantization, I lose almost all detection accuracy — either I get zero detections or very low confidence scores.

**Comment by pderrenger** [comment]
That symptom (zero/very low-confidence detections) is usually caused by either calibrating INT8 with a non-representative (or missing) dataset during export, or feeding the quantized TFLite model the wrong input scale/zero-point at runtime (e.g., still doing float 0–1 normalization instead of using the model’s quantized input format); can you share your exact `export.py` command and confirm the `.tflite` runs correctly with YOLOv5 `detect.py`/`val.py` on your desktop before compiling for Coral? See the [Ultralytics Docs on export](https://docs.ultralytics.com/modes/export/) and the [Ultralytics guide to model quantization](https://www.ultralytics.com/glossary/model-quantization) for the calibration/runtime details.

**Comment by muetazabdulsamad-oss** [comment]
Could you please provide the steps to perform this process?
Additionally, how can I calculate the mAP after converting the model to
Full INT8 quantization?

في الخميس، ٢٦ فبراير ٢٠٢٦, ١٢:٤٨ م Paula Derrenger ***@***.***>
كتب:

**Comment by pderrenger** [comment]
To get a fully INT8 TFLite and measure post-quantization accuracy, export with calibration data then validate that `.tflite` directly: `python export.py --weights best.pt --include tflite --int8 --data data.yaml --img 640 --batch-size 1`, then compute mAP with `python val.py --weights best-int8.tflite --data data.yaml --img 640 --conf 0.001 --iou 0.65` (this prints COCO-style mAP and saves outputs under `runs/val/`); once that looks good, compile for Coral with `edgetpu_compiler best-int8.tflite` (or export with EdgeTPU included) and treat the `best-int8.tflite` mAP as your accuracy reference. See the [YOLOv5 export tutorial](https://github.com/ultralytics/yolov5/issues/251) for the full set of export flags.
